File: LHDOntologyCleanup/src/main/java/cz/vse/lhd/lhdontologycleanup/OntologyMapping.java
package cz.vse.lhd.lhdontologycleanup;

import java.io.BufferedReader;
import java.io.DataInputStream;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 *
 * @author tomas
 */
public class OntologyMapping {

    private HashMap<String, ArrayList<String>> nameClassMappings = new HashMap();
    private final static String subclassPredicate = "http://www.w3.org/2000/01/rdf-schema#subClassOf";
    private String lang;

    public OntologyMapping(String ontologyPath, String supplementalOntologyPath, String _lang) throws IOException {
        lang = _lang;
        readOntologyTypes(ontologyPath, lang);
        //experimental -  does not work
        readSupplementalOntologyTypes(supplementalOntologyPath);

    }

    private String getDBpediaBaseURI() {

        if (lang.equals("en")) {
            return "http://dbpedia.org/resource/";
        } else {
            return "http://" + lang + ".dbpedia.org/resource/";
        }
    }

    public String map_sameAs(String uri) {
        String name = uri.substring(getDBpediaBaseURI().length()).toLowerCase();
        name = name.replace("_", " ");
        ArrayList<String> possibleClasses = nameClassMappings.get(name);

        if (possibleClasses == null) {
            return null;
        } else {
            return selectClass(name, possibleClasses);
        }
    }

    private String selectClass(String conceptName, ArrayList<String> possibleClasses) {
        ArrayList<String> uniqueClasses = new ArrayList();
        for (String possibleClass : possibleClasses) {
            if (!uniqueClasses.contains(possibleClass)) {
                uniqueClasses.add(possibleClass);
            }
        }
        if (uniqueClasses.size() == 1) {
            return uniqueClasses.get(0);
        } else {
            //TODO: prefer class from supplemental ontology
            System.out.println("Arbitrarily selecting from " + uniqueClasses.size() + " candidates");
            return uniqueClasses.get(0);
        }
    }
    /*
     * returns the most precise subclass  for argument
     * subclass is identified as class with longer name, containing the name of the class in the argument as substring
     * most precise = longest name, the subclass name must end with the argument
     */

    public String map_subclass(String uri) {
        String bestMatchingName = null;

        int bestMatchingNameLength = Integer.MIN_VALUE;
        String name = uri.substring(getDBpediaBaseURI().length()).toLowerCase();
        for (String candidateName : nameClassMappings.keySet()) {
            if (candidateName.matches(".*\\b" + name + "$") && candidateName.length() > name.length()) {
                int curLength = candidateName.length();
                /* rightmost matches are preferred
                 * e.g. in LawFirm  Firm is preferred before Law
                 */
                if (curLength > bestMatchingNameLength) {
                    bestMatchingName = candidateName;
                    bestMatchingNameLength = curLength;
                }
            }
        }
        if (bestMatchingName != null) {
            return selectClass(bestMatchingName, nameClassMappings.get(bestMatchingName));
        } else {
            return null;
        }
    }

    /*
     * returns the most precise superclass  for argument
     * superclass is identified as class with shorter name, contained in the name of the class in the argument as substring
     * most precise = longest of matching
     */
    public String map_superclass(String uri) {
        String bestMatchingName = null;

        int bestMatchingNameLength = Integer.MIN_VALUE;
        String name = uri.substring(getDBpediaBaseURI().length()).toLowerCase();
        for (String candidateName : nameClassMappings.keySet()) {
            if (name.matches(".*" + candidateName + "s?$") && candidateName.length() < name.length()) {
                int curLength = candidateName.length();
                /* rightmost matches are preferred
                 * e.g. in LawFirm  Firm is preferred before Law
                 */
                if (curLength > bestMatchingNameLength) {
                    bestMatchingName = candidateName;
                    bestMatchingNameLength = curLength;
                }
            }
        }
        if (bestMatchingName != null) {
            return selectClass(bestMatchingName, nameClassMappings.get(bestMatchingName));
        } else {
            return null;
        }
    }

    /*
     * Example input dbpedia_3.8.owl
     *    <owl:Class rdf:about="http://dbpedia.org/ontology/BasketballLeague">
     <rdfs:label xml:lang="en">basketball league</rdfs:label><rdfs:label xml:lang="el">Ομοσπονδία Καλαθοσφαίρισης</rdfs:label><rdfs:label xml:lang="fr">ligue de basketball</rdfs:label><rdfs:comment xml:lang="en">a group of sports teams that compete against each other in Basketball</rdfs:comment><rdfs:subClassOf rdf:resource="http://dbpedia.org/ontology/SportsLeague"></rdfs:subClassOf>
     </owl:Class><owl:Class rdf:about="http://dbpedia.org/ontology/LunarCrater">
     <rdfs:label xml:lang="en">lunar crater</rdfs:label><rdfs:label xml:lang="fr">cratère lunaire</rdfs:label><rdfs:label xml:lang="el">Σεληνιακός κρατήρας</rdfs:label><rdfs:subClassOf rdf:resource="http://dbpedia.org/ontology/NaturalPlace"></rdfs:subClassOf>
     </owl:Class>
     */
    private void readOntologyTypes(String ontologyPath, String lang) throws IOException {
        Pattern classPattern = Pattern.compile("<owl:Class rdf:about=\"([^\"]+)\">(.*?)</owl:Class>", Pattern.DOTALL);

        FileInputStream fstream = new FileInputStream(ontologyPath);
        // Get the object of DataInputStream
        DataInputStream in = new DataInputStream(fstream);
        BufferedReader br = new BufferedReader(new InputStreamReader(in));
        StringBuilder all = new StringBuilder();
        String thisLine;
        while ((thisLine = br.readLine()) != null) {
            all.append(thisLine);
        }
        Matcher classMatcher = classPattern.matcher(all);

        while (classMatcher.find()) {
            String className = classMatcher.group(1);
            //there are no alternative names for nl in dbpedia_3.8.owl, but many for de
            ArrayList<String> altNames = getAlternativeNames(classMatcher.group(2), lang);

            for (String altName : altNames) {
                ArrayList<String> nameClassMapping = nameClassMappings.get(altName);
                if (nameClassMapping == null) {
                    nameClassMapping = new ArrayList();
                    nameClassMapping.add(className);
                    nameClassMappings.put(altName, nameClassMapping);
                } else {
                    nameClassMapping.add(className);
                }
            }
        }
    }

    /*
     * Example input in nt syntax
     <http://dbpedia.org/resource/Display> <http://www.w3.org/2000/01/rdf-schema#subClassOf> <http://dbpedia.org/ontology/Play> .
     <http://dbpedia.org/resource/Boxers> <http://www.w3.org/2000/01/rdf-schema#subClassOf> <http://dbpedia.org/ontology/Boxer> .

     */
    private void readSupplementalOntologyTypes(String ontologyPath) {
        FileInputStream fstream = null;
        try {
            fstream = new FileInputStream(ontologyPath);
            // Get the object of DataInputStream
            DataInputStream in = new DataInputStream(fstream);
            BufferedReader br = new BufferedReader(new InputStreamReader(in));
            StringBuilder all = new StringBuilder();
            String thisLine;
            int lineCounter = 0;
            while ((thisLine = br.readLine()) != null) {
                lineCounter++;
                if (thisLine.startsWith("#")) {
                    continue;
                }
                if (!thisLine.contains(subclassPredicate)) {
                    continue;
                }
                int indexOfSubjectEnd = thisLine.indexOf(" ");
                String subject = thisLine.substring(1, indexOfSubjectEnd - 1);
                String subjectName = subject.substring(subject.lastIndexOf("/") + 1);
                /*                 int indexOfObjectNameStart = thisLine.lastIndexOf("/");//indexOfObjectEnd + predicateLength + 3;
                 int indexOfObjectStart = thisLine.lastIndexOf("<");
                 int indexOfObjectEnd = thisLine.lastIndexOf(">");
                 String objectName = thisLine.substring(indexOfObjectNameStart+1,indexOfObjectEnd);
                 String object = thisLine.substring(indexOfObjectStart+1,indexOfObjectEnd);*/

                ArrayList<String> nameClassMapping = nameClassMappings.get(subjectName);
                if (nameClassMapping == null) {
                    nameClassMapping = new ArrayList();
                    nameClassMapping.add(subject);
                    nameClassMappings.put(subjectName.toLowerCase(), nameClassMapping);
                } else {
                    System.out.println("Mapping for " + subjectName + " already loaded from the primary ontology");
                }

            }
        } catch (IOException ex) {
        } finally {
            try {
                if (fstream != null) {
                    fstream.close();
                }
            } catch (IOException ex) {
                Logger.getLogger(OntologyMapping.class.getName()).log(Level.SEVERE, null, ex);
            }
        }

    }

    /*
     * example input
     *  <rdfs:label xml:lang="en">basketball league</rdfs:label><rdfs:label xml:lang="el">Ομοσπονδία Καλαθοσφαίρισης</rdfs:label><rdfs:label xml:lang="fr">ligue de basketball</rdfs:label><rdfs:comment xml:lang="en">a group of sports teams that compete against each other in Basketball</rdfs:comment><rdfs:subClassOf rdf:resource="http://dbpedia.org/ontology/SportsLeague"></rdfs:subClassOf>
     */
    private ArrayList<String> getAlternativeNames(String classElemetContent, String lang) {
        Pattern alternativeNamesPattern = Pattern.compile("<rdfs:label xml:lang=\"" + lang + "\">(.*?)</rdfs:label>", Pattern.DOTALL);
        Matcher classMatcher = alternativeNamesPattern.matcher(classElemetContent);

        ArrayList<String> altNames = new ArrayList();
        while (classMatcher.find()) {
            altNames.add(classMatcher.group(1).toLowerCase());

        }
        return altNames;
    }
}

File: LHDOntologyCleanup/src/main/scala/cz/vse/lhd/lhdontologycleanup/OntologyMapping.scala
package cz.vse.lhd.lhdontologycleanup

import java.io.File
import java.util.regex.Pattern

import scala.annotation.tailrec
import scala.util.Try
import scala.xml.XML

/**
 * Created by propan on 18. 5. 2015.
 */
sealed trait OntologyMapping {
  def mapResourceToOntology(resource: String): Option[String]

  def mapResourceToOntologySubclass(resource: String): Option[String]

  def mapResourceToOntologySuperclass(resource: String): Option[String]
}

class SingleOntologyMapping(ontology: File, lang: String) extends OntologyMapping {

  private val nameClass = (XML.loadFile(ontology) \\ "Class")
    .filter(_.prefix == "owl")
    .flatMap { classElement =>
    val dbpediaClass = (classElement \ s"@{${classElement.getNamespace("rdf")}}about").text.trim
    val classLabels = (classElement \\ "label")
      .filter(labelElement => (labelElement \ s"@{${labelElement.getNamespace("xml")}}lang").text == lang)
      .map(_.text)
      .toSet
    val dbpediaClassName = dbpediaClass.replaceAll(Conf.dbpediaOntologyUri, "")
    (classLabels ++ Set(
      dbpediaClassName,
      dbpediaClassName.replaceAll("_", " "),
      dbpediaClassName.replaceAll("(?=\\p{Lu})", " ")
    )).map(_.trim.toLowerCase -> dbpediaClass)
  }.filter {
    case (x, y) => x.nonEmpty && y.nonEmpty
  }.toMap

  private def resourceName(resource: String) = resource.replaceAll(s"${Conf.dbpediaResourceUriRegexp}|${Conf.dbpediaOntologyUri}", "").replaceAll("_", " ").trim.toLowerCase

  def mapResourceToOntology(resource: String) = nameClass.get(resourceName(resource))

  /**
   * returns the most precise subclass for argument
   * subclass is identified as class with longer name, containing the name of the class in the argument as substring
   * most precise = longest name, the subclass name must end with the argument
   * @param resource dbpedia resource uri
   * @return
   */
  def mapResourceToOntologySubclass(resource: String) = {
    val matchRegexp = ".+\\b" + Pattern.quote(resourceName(resource))
    Try(
      nameClass.get(
        nameClass
          .keysIterator
          .filter(_.matches(matchRegexp))
          .maxBy(_.length)
      )
    ).getOrElse(None)
  }

  /**
   * returns the most precise superclass for argument
   * superclass is identified as class with shorter name, contained in the name of the class in the argument as substring
   * most precise = longest of matching
   * @param resource dbpedia resource uri
   * @return
   */
  def mapResourceToOntologySuperclass(resource: String) = Try(
    nameClass.get {
      val resourceNorm = resourceName(resource)
      nameClass
        .keysIterator
        .filter(ontologyClassName => resourceNorm.matches(".*" + Pattern.quote(ontologyClassName) + "s?") && resourceNorm.length > ontologyClassName.length)
        .maxBy(_.length)
    }
  ).getOrElse(None)

}

class MultipleOntologyMapping(implicit ontologyMapping: OntologyMapping*) extends OntologyMapping {

  @tailrec
  private def findFirst(f: OntologyMapping => Option[String])(implicit oms: Seq[OntologyMapping] = ontologyMapping): Option[String] = oms match {
    case om :: rest => f(om) match {
      case result @ Some(_) => result
      case None => findFirst(f)(rest)
    }
    case _ => None
  }

  def mapResourceToOntology(resource: String) = findFirst { om =>
    om.mapResourceToOntology(resource)
  }

  def mapResourceToOntologySubclass(resource: String) = findFirst { om =>
    om.mapResourceToOntologySubclass(resource)
  }

  def mapResourceToOntologySuperclass(resource: String) = findFirst { om =>
    om.mapResourceToOntologySuperclass(resource)
  }

}